support/ebpf/off_cpu.ebpf.c

```diff
@@ -0,0 +1,86 @@
+#include "bpfdefs.h"
+#include "types.h"
+#include "tracemgmt.h"
+
+// kprobe_progs maps from a program ID to a kprobe eBPF program
+bpf_map_def SEC("maps") kprobe_progs = {
+  .type = BPF_MAP_TYPE_PROG_ARRAY,
+  .key_size = sizeof(u32),
+  .value_size = sizeof(u32),
+  .max_entries = NUM_TRACER_PROGS,
+};
+
+// profile_off_cpu communicates scheduler tasks.
+bpf_map_def SEC("maps") profile_off_cpu = {
+  .type = BPF_MAP_TYPE_LRU_PERCPU_HASH,
+  .key_size = sizeof(u64),   // pid_tgid
+  .value_size = sizeof(u64), // time in ns
+  .max_entries = 256,
+};
+
+// tracepoint__sched_switch serves as entry point for off cpu profiling.
+SEC("tracepoint/sched/sched_switch")
+int tracepoint__sched_switch(void *ctx) {
+  u64 pid_tgid = bpf_get_current_pid_tgid();
+  u32 pid = pid_tgid >> 32;
+  u32 tid = pid_tgid & 0xFFFFFFFF;
+
+  if (pid == 0 || tid == 0) {
+    return 0;
+  }
+
+  u32 key = 0;
+  SystemConfig* syscfg = bpf_map_lookup_elem(&system_config, &key);
+  if (!syscfg) {
+    // Unreachable: array maps are always fully initialized.
+    return ERR_UNREACHABLE;
+  }
+
+  if (bpf_get_prandom_u32()%OFF_CPU_THRESHOLD_MAX > syscfg->off_cpu_threshold) {
+    return 0;
+  }
```

Comment:
I'm not sure the numbers will be statistically valid given this check. Shouldn't the probability of sampling be based on the amount of time spent? (Similar to how e.g. the probability of heap allocation samplers recording a stack is determined by the amount of bytes allocated).

Imagine if the off-CPU time of some process is dominated by one call to `read` that hangs forever (e.g. due to networking misconfiguration). Unless you get lucky enough that you hit this threshold on that call, you will not see it reflected in the profile.

Reply:
> Imagine if the off-CPU time of some process is dominated by one call to read that hangs forever (e.g. due to networking misconfiguration). Unless you get lucky enough that you hit this threshold on that call, you will not see it reflected in the profile.

I think you describe a good example of the general risk of a sampling approach. In https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/144 a sampling approach is proposed, as handling every scheduling event in both hooks is a too high risk to overload the system. If the sampling should be done on the off-CPU time, then there is also a management overhead to correctly size the eBPF map that communicates the start of the off-CPU time to the end of the off-CPU time. So yes, using a sampling approach to drop some events in the first hook will miss out on some events.
But similar to regular sampling based on-CPU profiling, if something is misconfigured, then there will be not a single event but multiple ones and so the issue will be visible to the profiler. The same applies to off-CPU profiling, I think. Overall, sampling based profiling provides valuable insights into the hot paths of the system for (expected or unexpected) things that happen often. Sampling is not an approach that catches every event, otherwise it would qualify as debugging or security utility - but sampling does not satisfy this requirements.

Reply:
I completely agree that the feature must use sampling. However the strategy used for sampling makes a difference.

E.g., see here where a similar analysis was done for jemalloc: https://github.com/jemalloc/jemalloc/blob/2a693b83d2d1631b6a856d178125e1c47c12add9/doc_internal/PROFILING_INTERNALS.md#L4

The authors concluded that they needed to sample per-byte, rather than per-allocation, because sampling per-allocation increases the variance significantly in a scenario where the allocations can have significantly different sizes.

The corresponding approach here would be to sample per-nanosecond, rather than per-event.

The downside is then you would have to call `bpf_ktime_get_ns` on the begin and end unconditionally (to know how long the task was parked for). I'm not sure what the overhead is.

I will try to put together a test based on your code to see how much this affects profiling variance in the real world.

Reply:
Thanks for the link to the research that got into the sampling strategy for jemalloc.

From my perspective, there are major differences in the purpose of sampling off CPU profiling vs jemalloc that lead to the differences in the sampling strategies.
jemalloc considers small allocations as “cheaper” whereas larger allocations result in more effort. For off CPU profiling the effort to take computing resources from a task and later reassign computing resources are always the same and the time a task is off CPU does not make a difference. So the sampling strategy of jemalloc takes the effort into account while for off CPU profiling the effort is constant.

The general event based sampling approach also comes with the advantage that it lets users predict the direct storage impact that off CPU profiling will have. The storage requirements are linearly correlated to configuration of the CLI flag `-off-cpu-threshold`. Changing the sampling strategy will make the storage predictions more complex and harder.

We can also not make assumptions about the environment, where off CPU profiling will be used. Will the environment be dominated by short interrupts, with small off CPU times for tasks, or will there be more larger off CPU times, e.g. if spinning disks are used instead of other faster memory solutions. The general event based sampling approach does not make a difference between these two kinds.


Switching to sampling on the off CPU time will also introduce a management overhead. First eBPF maps need to be scaled larger, resulting in a larger memory requirement for the eBPF Profiling agent and secondly there will be more computing overhead. To make a fair sampling decision on the off CPU time we need to make sure, that every tasked that is handled by the scheduler fits into the eBPF maps. The additional computing overhead will be on calling `bpf_ktime_get_ns` twice to get the off CPU time for each task before making a sampling decision on it, and more eBPF map updates and lookups to forward the off CPU time information from the first to the second hook. 

With https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/144 a general sampling approach was accepted, which is implemented in this current state of code proposal.
Maybe @open-telemetry/profiling-maintainers can provide their views on the different kinds of sampling strategies and provide guidance moving on on this topic.

Reply:
Thanks for the detailed response. Your points about the differences in constraints between jemalloc and opentelemetry-ebpf-profiler are compelling.

I still think in the future it is worth testing with different sampling strategies to see how much it affects variance in common scenarios and then making a call whether that's worth the downsides (e.g. less predictable storage usage).

But you've convinced me that this is a good enough default that it's worth starting with this.

Reply:
@umanwizard You bring up a valid point, in some tests I did for an unrelated project, calling `bpf_ktime_get_ns` too much can add up to significant overhead but that was at rates far higher than 500-1000Hz which I would assume is the worst-case of what we'd be dealing with here.

Reply:
> jemalloc considers small allocations as “cheaper” whereas larger allocations result in more effort.

I'll admit that I need to find some more time to fully digest the highly interesting jemalloc doc. But my initial impression is that this aspect was not the deciding factor for their choice of sampling strategy. I could be wrong tho.

> The storage requirements are linearly correlated to configuration of the CLI flag -off-cpu-threshold. Changing the sampling strategy will make the storage predictions more complex and harder.

Does it? Assuming we want to sample from `N` events that are produced of some time period `t`, a Bernoulli strategy using a time rate `R` will produce data at a rate of `t/R` (independent of `N`).

A strategy that relies on sampling 1 event for every `R` events will produce data at a rate of `N/R` (independent of `t`).

In both cases the choice of `R` allows linear control over the data rate, but the second strategy requires the user to know `N` in order to pick an initial rate. But `N` might change over time, so this strategy doesn't really allow a predictable data rate from a user perspective as far as I can tell.

> With https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/144 a general sampling approach was accepted, which is implemented in this current state of code proposal.

I think the merging of that PR signaled high-level consensus on implementing Off-CPU, but not necessary any consensus on the sampling strategy. I explicitly called that out in [my comment](https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/144#pullrequestreview-2276972138) here.

> Switching to sampling on the off CPU time will also introduce a management overhead.

I don't have immediate comments on this. It definitely needs to be considered. But I think there is also some value in thinking about the strategy we would _like_ if implementation issues were not prohibitive.

Reply:
I thought I replied to this thread at the beginning of the week. But it looks like I didn’t. So, sorry for the late reply. 

>> With #144 a general sampling approach was accepted, which is implemented in this current state of code proposal.
>
> I think the merging of that PR signaled high-level consensus on implementing Off-CPU, but not necessary any consensus on the sampling strategy. I explicitly called that out in my comment here.

You are right, #144 did not focus on a sampling strategy. My motivation of #144 was to introduce the general idea of off-CPU profiling and keeping its overhead minimal, so that there is no performance impact.

I think there is a major difference between the sampling strategies, per-allocation and per-byte, as used for jemalloc and the strategies, event based and off-CPU-time based, discussed here for off-CPU profiling.
The sampling strategy for jemalloc comes with the major advantage that at the time of decision the number of requested bytes are known. So there is no management overhead in changing the sampling strategy.
Whereas for off-CPU profiling, the time a task will be off-CPU is not known at the point when a scheduler decides to take computing resources from a task. To guarantee a fair sampling strategy based on off-CPU-time, it needs to be guaranteed that all tasks are tracked. Which will result in significantly more eBPF map updates/lookups but also the eBPF maps need to be scaled (significantly larger and) correctly. Besides this, there will also be some additional overhead in calling helpers like bpf_ktime_get_ns more often than required.

> [..] But my point about the unpredictable data volumes for the strategy in this PR remains valid I think.

Can you elaborate a bit more on this? With the suggested event based sampling strategy, the storage requirement should be linear in-/decreasing depending on the following major components

- Number of CPU cores
- CPU frequency
- off-CPU threshold

As the number of CPU cores should be known, as well their CPU frequency, the required storage is depending directly on the configured off-CPU threshold and will increase and decrease accordingly. Of course, this is a simplified view as environments change, CPU cores use a dynamic frequency and power management can take down CPU cores. But these environmental beneficial features should only reduce the required storage and the configuration with the most impact should be the configured off-CPU threshold.

Reply:
friendly ping for feedback @felixge - or can this conversation be resolved?

Reply:
friendly ping for feedback @felixge - or can this conversation be resolved?

Reply:
Although the sampling issue has been discussed before, I must reiterate that the current head-based sampling method has some drawbacks. These include challenges in ensuring the integrity of the sampling chain and the occasional problem of high latency in user localization. A client-based tail sampling approach might be a better-balanced solution. You can refer to this paper for more details: [The Beneﬁt of Hindsight: Tracing Edge-Cases in Distributed Systems](https://arxiv.org/pdf/2202.05769).

Reply:
[The Benefit of Hindsight](https://arxiv.org/pdf/2202.05769) is definitely an interesting read on retroactive sampling in distributed systems. But still I believe we should proceed with the current head sampling approach in this PR.

The primary focus here is implementing off-CPU profiling as a technique and functionality also enabling other use cases like on-demand profiling data using kprobes, and the current implementation offers practical benefits:

- Simpler visualization and understanding of the sampling data
- Lower CPU overhead (personal benchmarks showed retroactive sampling increasing CPU usage by ~8% and memory footprint by ~500MB)
- More straightforward implementation with less complexity

While retroactive sampling has its merits, defining precise conditions for it in advance is challenging. The current head sampling provides a general approach and additional filtering during data ingestion is still possible.

Rather than blocking this PR on sampling strategy discussions, I suggest we:

Move forward with the current head sampling implementation and consider sampling improvements and experiments as follow-up enhancements. This way we can get the core off-CPU profiling functionality in place while keeping the door open for future sampling strategy optimizations. Thoughts?

Reply:
I support moving forward with this approach. We can add additional off-CPU strategies in the future. Since every strategy will have its pros and cons depending on the context, it is IMO more important to move on and debate async on different strategies.
Additional future strategies for collecting off-CPU data allows users to decide which one is best for their use case.